Adding map name support to parameter Launch.Map=<map-name-or-uid>

AspectInteractive2 opened the pull request:
This change adds support for using relative paths with forward slashes in the Launch.Map=<map-path-or-uid> parameter, when launching the game via command line, by re-using the code from utility.cmd, which currently allows such paths. Previously if not using a UID, only an absolute path was allowed, such as shown in the below commands:

launch-game.cmd Game.Mod=ra **Launch.Map="C:\OpenRA\mods\ra\maps\a-nuclear-winter"**
launch-game.cmd Game.Mod=ra **Launch.Map=C:\OpenRA\mods\ra\maps\warwind.oramap**

Now these maps may also be loaded using the below commands:

launch-game.cmd Game.Mod=ra **Launch.Map=mods/ra/maps/a-nuclear-winter**
launch-game.cmd Game.Mod=ra **Launch.Map="mods/ra/maps/warwind.oramap"**

Note that quotes are not needed but allowed, as can be seen here. Both unpacked maps (map folders) and packed maps (.oramap files) work with this change. Also note that absolute paths such as those from the first example still work with this change. Have also tested running the game without the Launch.Map parameter to check for any regression errors in single player, and no issues have arose.

anvilvapre commented on OpenRA.Game/Game.cs:
```diff
@@ -956,9 +956,8 @@ public static void LoadMap(string launchMap)
-			var map = ModData.MapCache.SingleOrDefault(m => m.Uid == launchMap) ??
-				ModData.MapCache.SingleOrDefault(m => m.Package.Name == path);
+			var package = new OpenRA.FileSystem.Folder(Platform.EngineDir).OpenPackage(launchMap, ModData.ModFiles);
+			var map = ModData.MapCache.SingleOrDefault(m => m.Uid == launchMap || m.Uid == Map.ComputeUID(package));
```
Suggestion to only call `OpenPackage` if `m.Uid != launchMap`. Opening and parsing a package is 'expensive'.

`OpenPackage` may return null. You are not explicitly considering this. `ComputeUID` is then called with `null` which it does not support.

abcdefg30 replied:
This is only called once during startup so unless there's an elegant way to parse the package only after the UID was not found I wouldn't worry about changing this for performance reasons. An idea might be to check if `launchMap` is an UID or path (via regex or something like that).

anvilvapre replied:
Or simpler, just run the old code first - if that fails try it as a filename.

AspectInteractive2 replied:
Yes that's why it checks for a UID before it checks for a path. If it isn't a UID or path then the parameter is invalid and it won't work regardless. Wanted to avoid regex since a hash could theoretically have any characters in it (ours is only numbers and letters, but that might change in the future). I could put it in a try/catch method though if you want it to be cleaner.

AspectInteractive2 replied:
I have updated the code to include a try/catch block for the NullReferenceException that occurs if no path can be found, and to exit with a message stating the path was invalid if no valid path was provided. This also works if the UID is invalid, since an invalid UID param defaults to checking if the param is a the path, and a UID will never resolve to a valid path, so it will also exit with the same message (have tested this).

anvilvapre commented on OpenRA.Launcher/Properties/launchSettings.json:
```diff
@@ -15,6 +15,10 @@
       "commandName": "Project",
       "commandLineArgs": "Engine.EngineDir=\"..\" Game.Mod=ts"
+    },
+    "Red Alert - Test Map": {
```
Was this meant to be committed, or did you use it for local testing?

AspectInteractive2 replied:
Yes only for testing, I'll remove it

AspectInteractive2 replied:
Have now removed this as it was only for testing purposes.

Mailaender commented on OpenRA.Game/Game.cs:
```diff
@@ -955,15 +955,18 @@ public static void LoadMap(string launchMap)
+			{
+				Console.WriteLine($"Invalid Launch.Map parameter given. No map found at {launchMap}. Exiting program.");
+				Exit();
+			}
```
This `try catch log` pattern is only very seldom used in OpenRA and not for trivial stuff like a null reference exception that should certainly be handled. This seems like a needless change, but if you insist, I suggest:

```csharp
throw new ArgumentException($"Could not find map '{launchMap}'.");
```

If the game does not start, you certainly shouldn't `Exit()` without an exit code that indicates the error.

abcdefg30 replied:
I don't think we should use try-catch here. What is the problem with checking `package` for `null` before computing the id?

AspectInteractive2 replied:
Hi @Mailaender , @abcdefg30 , thanks for the feedback, have replaced the try/catch with a 'throw new ArgumentException', by checking whether the package variable is null before attempting to use ComputeUID(), as was recommended by abcdefg30. I can confirm that with this change ComputeUID no longer results in a NullPointerException, while still exiting out via the aforementioned throw exception if an invalid map param was provided. Have also tested again loading maps with a UID, absolute path, and relative path, and all are working as before. Please let me know if you need any further changes to the code.

pchote commented:
There is a major problem here that nobody has raised yet: this PR is all about loading maps from paths, but the game will fail with a bogus error message unless you give a path to the system or user map directory. Given this constraint, why bother with the path at all? The filename would be enough.

Usecase: I download a map from a bug report (e.g. #19579) and want to be able to launch it directly from my Downloads folder instead of installing it properly.

```
paul@Highwind:OpenRA+pr/19602 $ ./launch-game.sh Game.Mod=ra Launch.Map=../../Downloads/testcrash2.oramap 
[usual startup output removed]
Invalid Launch.Map parameter given. No map found at ../../Downloads/testcrash2.oramap. Exiting program.
```

But the map does indeed exist at that path:
```
paul@Highwind:OpenRA+pr/19602 $ ls ../../Downloads/testcrash2.oramap
../../Downloads/testcrash2.oramap
```

AspectInteractive committed:
Added map name support to Launch.Map command line parameter

Fixed code based on feedback

Replaced try/catch block with a null check and exception throw

Fixed code based on feedback

Fixed code based on feedback

Simplified Launch.Map parameter to use map name directly

AspectInteractive2 commented:
> There is a major problem here that nobody has raised yet: this PR is all about loading maps from paths, but the game will fail with a bogus error message unless you give a path to the system or user map directory. Given this constraint, why bother with the path at all? The filename would be enough.
> 
> Usecase: I download a map from a bug report (e.g. #19579) and want to be able to launch it directly from my Downloads folder instead of installing it properly.
> 
> ```
> paul@Highwind:OpenRA+pr/19602 $ ./launch-game.sh Game.Mod=ra Launch.Map=../../Downloads/testcrash2.oramap 
> [usual startup output removed]
> Invalid Launch.Map parameter given. No map found at ../../Downloads/testcrash2.oramap. Exiting program.
> ```
> 
> But the map does indeed exist at that path:
> 
> ```
> paul@Highwind:OpenRA+pr/19602 $ ls ../../Downloads/testcrash2.oramap
> ../../Downloads/testcrash2.oramap
> ```

Yes this seems to be a problem with the engine itself, rather than the path, as even loading it from a local folder that's outside the maps folder fails.

It may be possible to allow it to use all three options: UID, path, filename, though that would take significant rework which may not be worth it in the end (beyond what has already been achieved here). I think it would be better to at least have this change implemented, and then the filename feature could be added at a later date if there is an appetite for it.

pchote commented:
Yes exactly, the current implementation is limited to selecting maps out of the MapCache. My argument is that under this limitation the concept of giving a path *at all* is bogus. That bogosity was already there, but I think it would be a mistake to embed it further.

A much simpler and less misleading approach is to change the existing code to use `ModData.MapCache.SingleOrDefault(m => Path.GetFileName(m.Package.Name) == launchMap)` and then launch the game using `Launch.Map=warwind.oramap`.

AspectInteractive2 commented:
> Yes exactly, the current implementation is limited to selecting maps out of the MapCache. My argument is that under this limitation the concept of giving a path _at all_ is bogus. That bogosity was already there, but I think it would be a mistake to embed it further.
> 
> A much simpler and less misleading approach is to change the existing code to use `ModData.MapCache.SingleOrDefault(m => Path.GetFileName(m.Package.Name) == launchMap)` and then launch the game using `Launch.Map=warwind.oramap`.

That makes sense, yes, the path would then be superfluous if it's not able to be used. I have adjusted the code so that it uses the snippet you pasted here, and can confirm it works with just the filename provided, e.g. Launch.Map=warwind.oramap or Launch.Map=a-nuclear-winter . Also tested with UIDs again and it works, as well as an invalid path/UID returning the same error.

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/3520783d4d328108d717b09452835002d537c2fb...2b18eedb0e1a9b65e6c1697c2b5c7b7ea23d17c9)

Mailaender commented:
First pull request merged. :confetti_ball:
